### DISCUSSION TITLE
How to get the best of declarative and imperative paradigms

### DISCUSSION THREAD

**Discussion by rcoreilly** [cogentcore/core#961] [Ideas]
# The Declarative Paradigm

Many standard GUI frameworks use the **declarative** paradigm, where you construct a new composite object that represents the desired state of a widget, and the framework figures out from that what needs to be updated relative to the current state (i.e., basically performing a `diff` and applying the resulting `patch` to the current state).  This is done  _every single time the state changes_!

The primary advantage of this paradigm is that the user doesn't need to worry about all the possible ways something might change : you can just create the desired state.

The primary disadvantage is that it ends up putting large chunks of GUI logic into _single huge monolithic functions_  that are deeply nested and complex, with potentially significant  amounts of embedded logic.  In Go terms, it is basically writing your entire GUI logic within a struct literal expression, which are arguably the most unpleasant and unwieldy expressions in the language.

These declarative constructor expressions need to be massively nested because otherwise you would have to deal with complex logic of finding a particular place to insert a new element, and dealing with the state updating logic that you're trying to avoid in the first place.  This means that you cannot use standard programming logic like `if` and `for` to flexibly build your GUI elements.

For example, here is an example from the main [Flutter sample app](https://github.com/flutter/samples/blob/main/material_3_demo/lib/home.dart):

```dart
Widget build(BuildContext context) {
    final screenHeight = MediaQuery.of(context).size.height;
    final trailingActionsBody = Container(
      constraints: const BoxConstraints.tightFor(width: 250),
      padding: const EdgeInsets.symmetric(horizontal: 30),
      child: Column(
        mainAxisAlignment: MainAxisAlignment.end,
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          Row(
            children: [
              const Text('Brightness'),
              Expanded(child: Container()),
              Switch(
                  value: useLightMode,
                  onChanged: (value) {
                    handleBrightnessChange(value);
                  })
            ],
          ),
          Row(
            children: [
              useMaterial3
                  ? const Text('Material 3')
                  : const Text('Material 2'),
              Expanded(child: Container()),
              Switch(
                  value: useMaterial3,
                  onChanged: (_) {
                    handleMaterialVersionChange();
                  })
            ],
          ),
          const Divider(),
          _ExpandedColorSeedAction(
            handleColorSelect: handleColorSelect,
            colorSelected: colorSelected,
            colorSelectionMethod: colorSelectionMethod,
          ),
          const Divider(),
          _ExpandedImageColorAction(
            handleImageSelect: handleImageSelect,
            imageSelected: imageSelected,
            colorSelectionMethod: colorSelectionMethod,
          ),
        ],
      ),
    );
    return screenHeight > 740
        ? trailingActionsBody
        : SingleChildScrollView(child: trailingActionsBody);
  }
```

The code ends up being full of inline ternary logic expressions and other ways of replacing the standard procedural / imperative programming control flow constructs.  The excessive boilerplate elements of the Dart language are not doing this code any favors, but even within Go, the equivalent declarative constructors end up being similarly horrible, here in the case of [go-app](https://github.com/maxence-charriere/lofimusic/blob/master/bin/lofimusic/radio.go):

```go
func (r *radio) Render() app.UI {
	return app.Main().
		Class("radio").
		Class("fill").
		Body(
			newYouTubePlayer().
				Class("radio-player").
				Class("fill").
				Radio(r.current).
				OnPlaybackChange(r.onPlaybackChange),
			ui.Shell().
				Class("radio-shell").
				Class("fill").
				PaneWidth(menuWidth).
				Menu(newNav().
					LiveRadios(r.lives).
					CurrentRadio(r.current)).
				HamburgerMenu(newNav().
					Class("background-overlay").
					LiveRadios(r.lives).
					CurrentRadio(r.current)).
				Content(
					app.Aside().
						Class("radio-update").
						Class("app-title").
						Class("hspace-out").
						Body(
							ui.Stack().
								Class("fill").
								Right().
								Middle().
								Content(
									app.If(r.isUpdateAvailable,
										newLink().
											Class("link-update").
											Class("glow").
											Label("Update").
											Icon(newSVGIcon().RawSVG(downloadSVG)).
											OnClick(r.onUpdateClick),
									),
								),
						),
					app.Div().
						Class("hspace-out").
						Class("vspace-content").
						Body(
							newInfo().
								Radio(r.current).
								Playing(r.isPlaying),
						),
				),
		)
}
```

Note the use of `app.If` to implement conditional logic, and there are similar `app.Range` constructs.

[Fyne](https://github.com/fyne-io/fyne/blob/master/cmd/fyne_demo/tutorials/dialog.go) uses an imperative mode of updating, but ends up making large nested structures like the declarative paradigm for the initial configuration, as in this example from their tutorial, where the `container.NewVBox` function takes a list of widgets to add:

```go
	return container.NewVScroll(container.NewVBox(
		widget.NewButton("Info", func() {
			dialog.ShowInformation("Information", "You should know this thing...", win)
		}),
		widget.NewButton("Error", func() {
			err := errors.New("a dummy error message")
			dialog.ShowError(err, win)
		}),
		widget.NewButton("Confirm", func() {
			cnf := dialog.NewConfirm("Confirmation", "Are you enjoying this demo?", confirmCallback, win)
			cnf.SetDismissText("Nah")
			cnf.SetConfirmText("Oh Yes!")
			cnf.Show()
		}),
		openFile,
		saveFile,
		openFolder,
		widget.NewButton("Color Picker", func() {
			picker := dialog.NewColorPicker("Pick a Color", "What is your favorite color?", func(c color.Color) {
				colorPicked(c, win)
			}, win)
			picker.Show()
		}),
		widget.NewButton("Advanced Color Picker", func() {
			advancedPicker.Show()
		}),
		widget.NewButton("Form Dialog (Login Form)", func() {
			username := widget.NewEntry()
			username.Validator = validation.NewRegexp(`^[A-Za-z0-9_-]+$`, "username can only contain letters, numbers, '_', and '-'")
			password := widget.NewPasswordEntry()
			password.Validator = validation.NewRegexp(`^[A-Za-z0-9_-]+$`, "password can only contain letters, numbers, '_', and '-'")
			remember := false
			items := []*widget.FormItem{
				widget.NewFormItem("Username", username),
				widget.NewFormItem("Password", password),
				widget.NewFormItem("Remember me", widget.NewCheck("", func(checked bool) {
					remember = checked
				})),
			}

			dialog.ShowForm("Login...", "Log In", "Cancel", items, func(b bool) {
				if !b {
					return
				}
				var rememberText string
				if remember {
					rememberText = "and remember this login"
				}

				log.Println("Please Authenticate", username.Text, password.Text, rememberText)
			}, win)
		}),
	))
```

Clearly, there is room for improvement here.  There are ways of breaking these functions into smaller components, but the basic issues remain.

Another critical issue with the declarative paradigm is that the final resulting state can end up being a random mashup of old and new, updated elements, which can create problems with coordinating across the different parts of the overall state.  This typically requires additional complexity and logic to distinguish between the "state" that might be overwritten by updates.

# The Imperative Paradigm

In the imperative paradigm, one calls functions to directly construct the GUI structure, so that standard control flow logic can be used.  For example, here's a sample from the Cogent Core demo:

```Go
func text(ts *core.Tabs) {
	tab := ts.NewTab("Text")

	core.NewText(tab).SetType(core.TextHeadlineLarge).SetText("Text")
	core.NewText(tab).SetText("Cogent Core provides fully customizable text elements that can be styled in any way you want. Also, there are pre-configured style types for text that allow you to easily create common text types.")

	for _, typ := range core.TextTypesValues() {
		s := strcase.ToSentence(typ.String())
		core.NewText(tab).SetType(typ).SetText(s)
	}
}
```

The `NewX` constructor functions take a parent `Widget` and automatically add themselves to the parent's list of children.

This results in simple, logical code, but what happens if you want to _update_ the configuration based on some kind of state change?

Simple updates to _state_, such as hover or changing text etc can all be managed by changing state of individual widgets, and are straightforward to handle.  Typically, widgets handle these updates themselves.  But what happens if you want to make changes across different levels of a set of nested widgets?  This logic can become a bit complex, in terms of finding and updating the specific widgets in question.  And this update logic ends up being physically separated from the original configuration code, so it is harder to keep the two synchronized.

Furthermore if you want to change the overall _structure_ of a set of nested widgets, it can become even more complicated.  Often you end up just resetting everything and rebuilding from scratch, which is obviously less efficient than a declarative mode "diff" update.

# The Best of Both Worlds

To get the best of both worlds, we need the following:
* Ability to use standard programming logic, and avoidance of massive nested constructors.
* Co-location of configuration and updating logic.
* Diff-like efficient updating of structure.
* Easy extensibility so end-users and composite elements can selectively override default configuration and updating logic.

**Comment by rcoreilly** [comment]
# Proposal: Hierarchical Config

This is a proposal to achieve the best of both worlds, based on the following elements.

## Diff-based `update` logic

The `base/update` package provides the diff-like updating logic in a generic form, making minimal "edits" to a slice needed to make it align with a target specification.  A key simplification is that the target "template" for what should be there is specified entirely in terms of a `string` name (which can be any kind of unique id, e.g., a hash code).  Existing items in the slice are efficiently moved into the right positions, no-longer-needed elements are deleted, and if any new items are needed, a `New` function is called with the target name and index to make it.

## Widget `Config` template

Every Widget will have a `Config` that provides a declarative-like template for its structure and the updating of its components, that then drives the `update` diff logic.

A key feature of this proposal is that we use a path-based string id instead of literally constructing all the widgets -- this keeps the structure flat and flexible.

This config must be regenerated every time a relevant state update takes place.  Unlike in the full declarative case, the template is very lightweight and it is not at all a performance issue to construct it repeatedly.

Timing logic:
* `OnInit` function sets the individual style and event handlers only for the widget itself.
* `Config` function creates the widgets own default config and applies it.  This happens every Update.  
* If we want to be able to override this default config, then we need an additional map of overrides with a stack of functions, which are applied after the native default config.

```go
// ConfigItem represents a Widget configuration element,
// with one New function to create and configure a new child Widget,
// and an Update function to update the state of that child Widget.
// Contains a list of Config for its children, which is all sorted
// out as items are added.
type ConfigItem struct {
	// Path is the / delimited path to the element
	Path string

	// New returns a new Widget of the correct type for this element,
	// fully configured and ready for use.
	New func() Widget

	// Update updates the widget based on current state, so that it
	// propertly renders the correct information.
	Update func(w Widget)

	// Config for Children elements.
	Children Config
}

// Config is an ordered list of [ConfigItem]s,
// ordered in progressive hierarchical order
// so that parents are listed before any children.
// The Key of the map is the path of the element.
type Config []*ConfigItem

// Add adds given config.  This should be called on the root level Config
// list.  Any items with nested paths are added to Children lists as
// appropriate.
func (c *Config) Add(path string, nw func() Widget, updt func(w Widget)) {
...
}
```

The resulting Config code for a `Button` looks like this, which is basically a repackaging of the current OnWidgetAdded and Config logic.  One clear benefit is putting everything in the same place.

```go
func (bt *Button) Config() {
	var config Config
	// we check if the icons are unset, not if they are nil, so
	// that people can manually set it to [icons.None]
	if bt.HasMenu() {
		if bt.Type == ButtonMenu {
			if bt.Indicator == "" {
				bt.Indicator = icons.KeyboardArrowRight
			}
		} else if bt.Text != "" {
			if bt.Indicator == "" {
				bt.Indicator = icons.KeyboardArrowDown
			}
		} else {
			if bt.Icon == "" {
				bt.Icon = icons.Menu
			}
		}
	}

	config.Add("parts", func() Widget {
		w := NewParts()
		w.Style(func(s *styles.Style) {
			s.Gap.Zero()
			s.Align.Content = styles.Center
			s.Align.Items = styles.Center
		})
		return w
	}, nil)

	if bt.Icon.IsSet() {
		config.Add("parts/icon", func() Widget { return NewIcon() },
			func(w Widget) { w.(*Icon).SetIcon(bt.Icon).Update() })
		if bt.Text != "" {
			config.Add("parts/space", func() Widget { return NewSpace() }, nil)
		}
	}
	if bt.Text != "" {
		config.Add("parts/text", func() Widget {
			w := NewText()
			if bt.Type == ButtonMenu {
				w.Type = TextBodyMedium
			} else {
				w.Type = TextLabelLarge
			}
			w.Style(func(s *styles.Style) {
				s.SetNonSelectable()
				s.SetTextWrap(false)
				s.FillMargin = false
			})
			return w
		}, func(w Widget) {
			w.(*Text).SetText(bt.Text).Update()
		})
	}
	...
	config.ConfigWidget(bt, "")
}
```

**Comment by rcoreilly** [comment]
# Improved semantic terminology

The process of constructing the overall GUI is composed of two processes, analogous to how almost every other complex entity is made.

1. First you `Make` the overall `Plan` that specifies how everything should be built, in a simple, direct, imperative manner, using function closures that directly style and configure a Widget element.  Each element is `Add`ed to the Plan, either with automatic default naming or `AddAt` a particular named path in the overall tree.

2. Then you `Build` the actual Widgets based on the Plan.  Critically, the Plan can be updated and modified dynamically as things change, and through the magic of software and the diff-based updating process, the resulting Build can dynamically change as well.

Keeping these two processes logically separate in your mind is key for understanding what goes where, including the critical point that the Plan can be continuously updated, but the actual built elements remain in place unless the plans for that element change.

Customization can happen at both stages of the process: the Plan can be updated by additional `Makers` that layer on top of the core default Make Plan process, and the `Build` function can be customized to perform additional steps prior to or after the `BuildFromPlan` function does the diff-based updating based on the plan.  The default `BuildWidget` wrapper method orchestrates these steps, calling Make to make the Plan, then calling Build, which just calls `BuildFromPlan` in its default `WidgetBase` version, but which can be modified to perform additional steps.

**Reply by rcoreilly** [reply]
* `MakeFuncs` is a stack of `Make(p *Plan)` functions that can be associated with any widget, especially frequently used for the overall `AppBar` `Toolbar`
* `BuildFuncs` is a stack of `Build(p *Plan )` functions that can also be layered on top of the default `Build` function, run after that function executes, to manage any post-build custom steps that might require coordination across different widgets -- these should be rare, and most of the building process should be specified locally within the Plan.

**Reply by rcoreilly** [reply]
In most cases, it is best to design in terms of components that you Plan and Build separately, rather than having everything be part of one giant Plan that encompasses everything in the app.  This allows flexibility in later redesigns, and simpler, more encapsulated elements that are then more robust and not so strongly interdependent on each other.

Any _necessary_ interdependency can hopefully be captured by layering `MakeFuncs` on relevant subcomponent widgets, to update their plans as needed.

**Comment by kkoreilly** [comment]
# Proposal: structurally nested widget planning

The key limitation of the existing proposal is that you can not make automatically named children, since you have to specify the nesting using paths, which are not possible with automatic naming. My proposal addresses this by using a similar system to the existing tree nesting paradigm we use.

In my proposal, you can write this:

```go
splits := core.Add(p, func(w *core.Splits) {})
core.Add(splits, func(w *core.Frame) {})
```

Instead of:

```go
core.AddAt(p, "splits", func(w *core.Splits) {})
core.AddAt(p, "splits/nav", func(w *core.Frame) {})
```

In my proposed code above, `splits` is a variable of type `*core.PlanItem`, and `core.Add` just nests appropriately in that, which should even already work without any additional code. I would even propose that we should stop supporting the slash based separation and use the much more robust and programmatic variable approach instead, which would also allow us to get rid of a lot of path handling code in config.

**Reply by rcoreilly** [reply]
Another option from Kai:

```go
core.Add(p, func(w *core.Splits) {
    w.AddMaker(func(p *core.Plan) {
        core.Add(p, func(w *core.Frame) {})
    })
})
```

**Reply by kkoreilly** [reply]
This is actually two tabs per level of nesting, which is quite extreme:

```go
core.Add(p, func(w *core.Frame) { // 1
	w.AddMaker(func(p *core.Plan) {
		core.Add(p, func(w *core.Frame) { // 2
			w.AddMaker(func(p *core.Plan) {
				core.Add(p, func(w *core.Frame) { // 3

				})
			})
		})
	})
})
```

**Reply by kkoreilly** [reply]
See fc9ff650dba547ece8942dbc7b6f9fa50ae806c1 for an example of the effects of this.

**Comment by kkoreilly** [comment]
# App-level declarative paradigm?

Here are two possible ways to write the basic example, both seven lines of code:

Existing way:

```go
package main

import "cogentcore.org/core/core"

func main() {
	b := core.NewBody("Hello")
	core.NewButton(b).SetText("Hello, World!")
	b.RunMainWindow()
}
```

Potential new declarative way:

```go
package main

import "cogentcore.org/core/core"

func main() { core.Run(NewApp()) }

type App struct { core.Frame }

func (a *App) Make(p *core.Plan) {
	core.Add(p, func(w *Button) { w.SetText("Hello, World!") })
}
```

**Comment by rcoreilly** [comment]
# Making plans and building widgets, with dynamic updating

(Note: This is provisional documentation for the current state of the new Plan / Build architecture.)

Perhaps the most important challenge for an advanced GUI framework is providing an efficient, logical, and flexible way to construct _and dynamically update_ the GUI contents.  One popular "brute force" technique, the _declarative_ approach, is to use exactly the same code for construction and updating, rebuilding the current state every time, and letting the framework figure out what has changed (using a diff-based algorithm).

This is not very computationally or memory efficient, to say the least: our benchmark tests showed factors in the 100x to 1000x range of both computation and memory cost relative to our more optimized approach.  It also typically makes it difficult or impossible to use standard programming logic during construction, because everything ends up being written in the form of a giant static literal expression.  Furthermore, it introduces significant complexity regarding the status of the active state of the widgets when they are updated.

Our "Plan & Build" alternative approach improves upon the standard declarative approach by using _unique name identifiers_ as _keys_ (like unique map keys) for determining the elements of a _Plan_ for the GUI structure, and putting all the construction code inside a _closure_ that is only executed when a new widget element is actually needed.  Thus, all the logic is expressed using standard programming constructs (`if`, `for` etc) and new elements are only built when needed.  Furthermore, the use of unique name id's allows the `diff`-like logic to efficiently _move_ items around for insertions and deletions, instead of triggering waves of deletions and construction of new items.

A key general principle behind the `core` design is to use stacks of closure functions to drive all the subsequent dynamic behavior and updating of the Widget over its subsequent lifetime.  Thus, it is important that all of this configuration happens only once when the Widget is first created, so there is one consistent stack of such closures: the Plan and Build structure is designed around this, with initialization functions that are run once when the widget is created, along with the ability to add additional closures that are run at the different levels of updating.  Once you have this picture of how things are configured, all of the logic should make good sense.

## `Build`, `Rebuild`, `Restyle`

At the broadest level, there are three main phases of construction and updating:

* Initial construction of the GUI for first presentation, which calls `Build` in the process of making the `Scene` ready to display.  This will trigger `OnInit` and `OnAdd` functions for each Widget, which are called when they are first created and added to their parent in the overall GUI tree structure (respectively).

* `Rebuild`: calls `Build` on an already-constructed GUI, when something structural might have changed (e.g., a Widget added or removed).  Because of the efficient Plan-based logic used for `Build`, this is very efficient in general.  It will trigger a `Layout` of the entire rendered `Scene`, which does add an additional time cost, but ensures that any structural changes are properly reflected in the layout of all the widgets.

* `Restyle`: is an even more efficient update specifically for `State` changes such as hovering or pressing a Button, enabling or disabling elements, etc.  This only triggers a `Render` of the widget in question, and skips the full layout, and does not make any structural changes.  The styling parameters are all designed so that standard state updates specifically do not require a new layout (e.g., the margin automatically accommodates all the possible border and shadow effects that might be triggered by different state-driven style updates).  This makes the overall GUI highly responsive and efficient even on slower platforms, for most standard user event-driven updates.  The end-user programmer typically does not need to manage this level of updating, as it is automatically handled by the standard Widgets.

## `Make` a `Plan`, `Build` Widgets

First, the overall _specification_ of how to make a widget is typically contained within its `OnInit` method, which is only called _once_ when the widget is first created (_prior_ to when it is added into the GUI tree).  Critically, this code is largely in the form of _closure_ functions that will be called _later_, e.g., during a `Build` call that happens initially when the `Scene` is first shown to the user.

A typical `OnInit` has 3 major sets of functions:

*  `Styler` calls for the widget itself, that set initial and updated styling, which are called during `Restyle` to update the widget rendering based on state changes.

* Event handling functions such as `OnClick` that determine what happens in response to user input or other events.

* `AddChild` functions that add `Make` functions for adding Widget elements to a `Plan`, thereby determining the structure of the widget when `Build` or `Rebuild` is called.

You can also optionally use `Builder` to add misc additional updating functions that are called during `Build`, outside of the `Plan` logic.

We will use a simple example of a widget with a `Text` label to illustrate:

```go
type MyWidget struct {
    core.Frame
    Text string
}

func (mw *MyWidget) OnInit() {
    mw.Frame.OnInit()
    // first, do any custom styling _of the widget itself_
    mw.Styler(func(s *styles.Style) {
        s.SetAbilities(true, abilities.Activatable)
        s.Border.Radius = styles.BorderRadiusFull
    })
    // then handle events
    mw.OnClick(func(e events.Event) {
        e.SetHandled()
        fmt.Println("clicked")
    })
    // Make a Plan for GUI structure
    core.AddChild(mw, func(w *core.Text) { // will be called with the newly made Text Widget
        w.Styler(func(s *styles.Style) { // optional extra styling for this widget
            s.Color = colors.C(colors.Red)
        })
        w.Builder(func() {         // will be called whenever Text widget is built
            w.SetText(mw.Text)  // this captures our state to update Text widget
        })
    })
}
```

As a result of this code, updating the text label can be triggered by:
```go
mw.SetText("new text").Rebuild()
```

The `core.AddChild` function combines these two steps:
```go
    mw.Maker(func(p *core.Plan) {
        core.Add(p, func(w *core.Text) {
```
Thus, it is adding a new Maker function to our Widget that will be called at `Build` time, and adding a `Plan` element to that function that will create a new widget of the given type (`core.Text`).  Because Go does not allow methods with their own separate generic types, we have to use global functions like `core.AddChild` to take advantage of the generic typing based on the call signature of the Add functions.

The plain `core.AddChild` function automatically generates a unique label for the Plan element based on the code filename, line number, and package, which should be sufficiently unique.  There is a `core.AddChildAt` version that takes an explicit name, which is useful if you need to later find that element by a path name, or you are creating elements in a `for` loop and thus need to give them unique names based on the loop counter.

After the initial `Build` call runs the code in our `AddChild` closure, that code will not be called again, because we don't have any conditional logic so the built structure will always match the Plan.  Note that the `Styler` closure will be called during `Restyle`.  However, the `Builder` closure _is always run at every Rebuild_, so the `SetText` call on the `Text` element defined there will be called (and Rebuild also rebuilds all of the widgets under the level at which it is called, including that Text element).  Note that all the `Styler` functions are also called during a `Rebuild`, along with `Layout`.

You should think of the Text widget as having children within it that represent the text elements that are rendered, so it needs to be rebuilt to update those children.  This is handled efficiently using a `render.Text` text layout instead of separate widgets, but this text layout is still a persistent layout of a given text string, which also drives the sizing of the Text element so it allocates enough room to render the text.

For plans that go beyond the immediate child widgets, you can continue to call `core.AddChild` function for _for that child widget_ that can add or modify the children of that widget, and so on.  For more deeply nested trees, it is a good idea to create `makeX(p *core.Plan)` methods that can be passed to `Maker` to handle the making of that sub-element.

**Comment by matta** [comment]
Disclaimer: I have only spent about an hour investigating cogentcore, and all of my GUI experience is from the 90s, before more modern trends became popular. The whole project strikes me as an interesting and impressive achievement. In particular, it is a GUI library that comes with multiple non-trivial applications, which surely provides invaluable experience. My comments here are merely first impressions.

Overall, I find the claims, made here and elsewhere in cogentcore documentation, against the declarative approach to be over-stated and under-substantiated. I read the claims and remain unconvinced. These concerns are closer to a matter of programmer taste and familiarity than a matter of being obvious or factual. Often declarative approaches are held up as being _easier_ to maintain precisely because they are a more constrained, less expressive, DSL, and so easier to learn and use correctly. This is in large part why declarative approaches like HTML have succeeded. Personally, I usually find nested declarations easier to reason about than a set of tightly woven imperative APIs.

That said, the efficiency arguments presented here hold obvious weight, in my opinion.

I noticed that the pathed ID approach talked about here bears some resemblance to the one used by Xilem, I presume for similar reasons. See "Identity and id paths" at https://raphlinus.github.io/rust/gui/2022/05/07/ui-architecture.html. That architecture sets the lofty goal of being both declarative *and* fast. But, it is written in Rust, and no Go solution is likely to be able to touch it in terms of performance.

> use stacks of closure functions

It isn't clear to me how these things are stacks. Stacks usually involve both push and pop operations, and often only the top element is accessed. Aren't these just lists?

After reading all the above posts I remained confused about the distinction between Planning, Building, Rebuilding, Restyling, etc. I *think* the planning phase is effectively constructing bespoke "objects" of what amounts to a single type that behaves in a fixed way according to "the plan". The other phases are effectively methods on that object type (or "class").

The final `MyWidget` example is an interesting one. The code is essentially building bespoke objects at run time, but through closures rather than the usual Go struct/method approach. My first reaction is that the resulting "closure soup" is quite confusing and likely to be difficult to maintain and debug for many people. Any approach that uses regular Go interfaces is likely to be preferable. I keep seeing a pattern in GUI frameworks written in languages without traditional OO -- they end up re-expressing standard OO patterns in bespoke ways.

**Reply by kkoreilly** [reply]
Thank you for your feedback!

First, I would like to note that this was a WIP discussion of how to best structure things, and thus some of the examples and principles stated here are not what actually ended up being implemented. For an accurate representation of the current paradigms, you can look at the following sections of our docs:

* The basics section, in particular everything between [widgets](https://www.cogentcore.org/core/basics/widgets) and [plans](https://www.cogentcore.org/core/basics/plans) (inclusive)
* The advanced section, in particular [styling](https://www.cogentcore.org/core/advanced/styling) and [updating](https://www.cogentcore.org/core/advanced/updating)

Nevertheless, this discussion still does capture a good amount of the overall structure and principles of our current paradigm, and your concerns are legitimate. I will do my best to address them here, and I would be glad to here your thoughts after reading those documentation pages and my points here. I am not ruling out any possibilities such as a declarative approach, just making my case.

> Overall, I find the claims, made here and elsewhere in cogentcore documentation, against the declarative approach to be over-stated and under-substantiated. I read the claims and remain unconvinced.

Let me first say that, relative to @rcoreilly, I was more in favor of a full declarative paradigm. I still agree with most of his arguments against a declarative paradigm, but I also recognize the merits of declarative paradigms to a further extent (having used them myself). I think that @rcoreilly's most adamant argument against declarative is the performance concerns, which I personally find to be somewhat overstated, so he would be better suited to make that case.

My biggest argument against a declarative approach is, as @rcoreilly said:

> The code ends up being full of inline ternary logic expressions and other ways of replacing the standard procedural / imperative programming control flow constructs.

As you can see in the code examples in the initial post above, all of the declarative paradigms use ternary expressions, `app.If`, or other such constructs that replace standard `if`, `for`, etc expressions found in idiomatic Go. This makes it much harder to neatly embed logic into GUI applications. For example, the `go-app` example in the original post doesn't really look like Go.

As you noted, we have built several large applications using Cogent Core. In those apps, we find that the amount of actual basic GUI code that would be easy with a declarative approach is small, and most of the code is focused on business logic, such as simulating neural networks and manipulating SVG vector graphics. What Cogent Core allows you to do is easily plug that imperative Go business logic into a GUI without worrying about state concerns and ternaries and other declarative issues.

> This is in large part why declarative approaches like HTML have succeeded.

The problem is that people are not satisfied with HTML, as evident by the constant creation of new JavaScript frameworks. In my opinion, the reason that people are not satisfied with HTML is because it separates GUI logic from business logic too much. This is the key advantage of our closure-based approach: you can very easily integrate logic into your GUI using idiomatic Go embedded in closures.

> It isn't clear to me how these things are stacks.

Yes, stacks was a misnomer, @rcoreilly meant lists. Thank you for pointing that out.

> After reading all the above posts I remained confused about the distinction between Planning, Building, Rebuilding, Restyling, etc.

If you are still confused about the distinction between all of those things after reading the documentation pages, please let me know and I will try to clarify it further (again, some concepts like Building do not exist in the current paradigm).

> My first reaction is that the resulting "closure soup" is quite confusing and likely to be difficult to maintain and debug for many people. Any approach that uses regular Go interfaces is likely to be preferable.

I guess I don't really understand what you mean by "regular Go interfaces" here, since we believe the advantage of our approach is that it is more like idiomatic Go than a declarative approach. For example, in my experience using go-app, I experienced an endless stream of hard-to-debug issues related to state management, component mounting, and control constructs.

Thank you again for your feedback!

**Reply by matta** [reply]
Ah, I had actually read through the docs and tutorials and thought this proposal was a modification the current release! Just goes to show how little I understood about both.

> As you can see in the code examples in the initial post above, all of the declarative paradigms use ternary expressions, app.If, or other such constructs that replace standard if, for, etc expressions found in idiomatic Go. This makes it much harder to neatly embed logic into GUI applications. For example, the go-app example in the original post doesn't really look like Go.

I don't really like things like `app.If` either.

> The problem is that people are not satisfied with HTML, as evident by the constant creation of new JavaScript frameworks. In my opinion, the reason that people are not satisfied with HTML is because it separates GUI logic from business logic too much. This is the key advantage of our closure-based approach: you can very easily integrate logic into your GUI using idiomatic Go embedded in closures.

Yes, HTML is not good at expressing highly dynamic and interactive applications.

I do notice that interactivity was initially added to HTML by attaching Javascript closures to DOM elements in a way that isn't all that different from the way Cogent Core wires up behavior using closures. People found the callback-heavy code hard to deal with, and so things like React came to be.

I don't have any answers here. I think GUI is a hard problem, and for each of the myriad of ways a GUI can be expressed one can find examples where that way works best.

> I guess I don't really understand what you mean by "regular Go interfaces" here.

Oh, for that I was thinking something like the below, but now I don't think this would fit well with Core's design as I understand it. It actually makes me a bit sad, because I've been looking for a way I can express a GUI directly in idiomatic code, of some simple language like Go, for a long time. Core comes close, and while it is true that it doesn't invent new language constructs, it does introduce a kind of bespoke object system constructed from lists of closures. Perhaps this sort of thing is unavoidable in the problem space.

```
type MyWidget struct {
	core.Frame
	Text string
}

func (mw *MyWidget) Styler() {
	s.SetAbilities(true, abilities.Activatable)
	s.Border.Radius = styles.BorderRadiusFull
}

func (my *MyWidget) OnClick(e events.Event) {
	e.SetHandled()
	fmt.Println("clicked")
}

func (my *MyWidget) Plan(....TBD....) {
	w.Styler(func(s *styles.Style) { // optional extra styling for this widget
		s.Color = colors.C(colors.Red)
	})
	w.Builder(func() { // will be called whenever Text widget is built
		w.SetText(mw.Text) // this captures our state to update Text widget
	})
}

func (mw *MyWidget) OnInit() {
	mw.Frame.OnInit()
}
```

**Reply by kkoreilly** [reply]
> I do notice that interactivity was initially added to HTML by attaching Javascript closures to DOM elements in a way that isn't all that different from the way Cogent Core wires up behavior using closures. People found the callback-heavy code hard to deal with, and so things like React came to be.

There are certainly some similarities, but a key feature of Cogent Core is that everything is in one cohesive language. One of the biggest issues with vanilla HTML/CSS/JS is that you have to constantly re-express how to get an element imperatively in different contexts, making it hard to track. For example, to make a red button that prints something when you click on it using separate (not inline) HTML/CSS/JS, you have to do something like this:

```html
<button id="my-button">Hello</button>
```
```css
#my-button {
  background-color: red;
}
```
```js
document.querySelector("#my-button").addEventListener("click", function (e) {
  console.log("Hello");
});
```

Notice how `my-button` is repeated three times, since you have to fetch/define the element in three different languages. With Cogent Core, you can do all of this in one language without having to retrieve anything:

```go
bt := core.NewButton(b).SetText("Hello")
bt.Styler(func(s *styles.Style) {
	s.Background = colors.Scheme.Error.Base
})
bt.OnClick(func(e events.Event) {
	fmt.Println("Hello")
})
```

Again, there are similarities between these approaches, but I would argue that one of the biggest advantages of React is that you don't have to separate HTML and JS; in fact, they even state so on their website:

> Putting JSX markup close to related rendering logic makes React components easy to create, maintain, and delete.

Cogent Core provides you the same advantage by coupling everything together instead of having separate things like HTML/CSS/JS, which I would argue is the main disadvantage of that approach. Furthermore, the [updaters](https://www.cogentcore.org/core/basics/updating) and [plans](https://www.cogentcore.org/core/basics/plans) in Cogent Core provide a lot of the same functionality as declarative approaches, allowing you to avoid the deeply nested imperative callbacks that plague JS.

> Oh, for that I was thinking something like the below, but now I don't think this would fit well with Core's design as I understand it.

We actually used to have a system a lot more like that, but the fundamental problem with such an approach is it makes it extremely difficult to customize widgets. For example, if you wanted to make a red button, you would have to make a whole new widget type since the only way to style something is to implement interface methods. If you wanted to make a button that prints something, you would also need to make a new widget type since the only way to add an event handler is to implement interface methods. That kind of ultra-idiomatic approach sounds great in theory, but as soon as you actually have to customize anything, it becomes a nightmare. In fact, other Go GUI frameworks such as Fyne and Gio take more of that approach, but then it is impossible to customize anything without making new custom widget types and writing a bunch of code. The lists of closures allow you to have easily customizable widgets without some separate language like CSS.

As you noted, this is a hard problem, and different approaches sacrifice different things. If you have any ideas on how we could ameliorate these problems without removing key functionality, we would be glad to consider them. Thank you again for all of your feedback!
